components/support/utils/src/main/java/mozilla/components/support/utils/observer/SideEffects.kt

```diff
@@ -0,0 +1,53 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.support.utils.observer
+
+/**
+ * This helper class prevents side effects (glitches) in an <code>Observable</code> implementation
+ * where an observer can modify the internal state of an <code>Observable</code> from an observer
+ * callback. In this situation follow-up observers may see a modified state that may conflict with
+ * the change the observer was notified about.
+ *
+ * To avoid those side effects in an <code>Observable</code> create an internal instance of
+ * <code>SideEffects</code> and wrap methods modifying the <code>Observable</code> state inside
+ * <code>modifyWithoutSideEffects()</code>. Methods that just read the current state can be wrapped
+ * in <code>readWithLock</code>.
+ *
+ * SideEffects queues modifications to an <code>Observable</code> until all observers have been
+ * notified and then performs them sequentially
+ *
+ * For an example of observer side effects / glitches see SideEffectsTest<code>.
+ */
+class SideEffects {
+    // Those properties have 'internal' visibility with @PublishedApi annotation so that they can be
+    // inlined by the compiler.
+
+    @PublishedApi internal val lock = Any()
+    @PublishedApi internal val queue: MutableList<() -> Unit> = mutableListOf()
+    @PublishedApi internal var isProcessing = false
+
+    inline fun modifyWithoutSideEffects(crossinline action: () -> Unit): Unit = synchronized(lock) {
+        if (isProcessing) {
+            queue.add {
+                action()
+            }
+        } else {
+            isProcessing = true
+
+            action()
```

Comment:
I'm afraid that a `return` statement inside the lambda will jump out of this method because it's getting inlined. I thought `crossinline` would prevent that. But I seem to be able to still use a return@label:

```
fun someMethod(bar: Foo?) = sideEffects.modifyWithoutSideEffects {
    if (bar == null) {
        // This doesn't compile because of 'crossinline'
        return
    }

    if (bar == null) {
        // This compiles... but isn't it the same as the one above?
        return@modifyWithoutSideEffects
    }

    // ...
}
```

Reply:
I wonder if every `action()` should actually be wrapped in:

```Kotlin
try {
    action()
} finally {
   processQueue()
}
```

... so that we will always continue with the next action if one of them returns after being inlined?

Reply:
This try catch doesn't really work... Every action we process in the queue would need to be wrapped in the same try-catch. We can do that with recursion - but we can't inline a method that is called recursively..

Reply:
Interesting. Shouldn't we just use noinline for the lambda here? It shouldn't have the power to jump out of the executing method :).

Reply:
I wanted to inline aggressively so that using SideEffects has no penalty. Otherwise Java (6) will need to create an object (+ the class) for every lambda :(